sentry/src/main/java/io/sentry/SentryTracer.java

```diff
@@ -52,29 +81,88 @@ public SentryTracer(
       final @NotNull IHub hub,
       final boolean waitForChildren,
       final @Nullable TransactionFinishedCallback transactionFinishedCallback) {
-    this(context, hub, null, waitForChildren, transactionFinishedCallback);
+    this(context, hub, null, waitForChildren, null, false, transactionFinishedCallback);
   }
 
   SentryTracer(
       final @NotNull TransactionContext context,
       final @NotNull IHub hub,
       final @Nullable Date startTimestamp) {
-    this(context, hub, startTimestamp, false, null);
+    this(context, hub, startTimestamp, false, null, false, null);
   }
 
   SentryTracer(
       final @NotNull TransactionContext context,
       final @NotNull IHub hub,
       final @Nullable Date startTimestamp,
       final boolean waitForChildren,
+      final @Nullable Long idleTimeout,
+      final boolean trimEnd,
       final @Nullable TransactionFinishedCallback transactionFinishedCallback) {
+    this(
+        context,
+        hub,
+        startTimestamp,
+        waitForChildren,
+        idleTimeout,
+        trimEnd,
+        transactionFinishedCallback,
+        null);
+  }
+
+  SentryTracer(
+      final @NotNull TransactionContext context,
+      final @NotNull IHub hub,
+      final @Nullable Date startTimestamp,
+      final boolean waitForChildren,
+      final @Nullable Long idleTimeout,
+      final boolean trimEnd,
+      final @Nullable TransactionFinishedCallback transactionFinishedCallback,
+      final @Nullable CountDownLatch latch) {
     Objects.requireNonNull(context, "context is required");
     Objects.requireNonNull(hub, "hub is required");
     this.root = new Span(context, this, hub, startTimestamp);
     this.name = context.getName();
     this.hub = hub;
     this.waitForChildren = waitForChildren;
+    this.idleTimeout = idleTimeout;
+    this.trimEnd = trimEnd;
     this.transactionFinishedCallback = transactionFinishedCallback;
+    this.latch = latch;
+
+    if (idleTimeout != null) {
+      scheduleFinish(idleTimeout, latch);
+    }
+  }
+
+  @TestOnly
+  void scheduleFinish(final @NotNull Long idleTimeout, final @Nullable CountDownLatch latch) {
+    cancelTimer();
+    timerTask =
+        new TimerTask() {
+          @Override
+          public void run() {
+            final SpanStatus status = getStatus();
+            finish((status != null) ? status : SpanStatus.OK);
+            if (latch != null) {
```

Comment:
I'm not sure if it's good idea to have this in prod code, but idk how to test it otherwise?

Reply:
You can implement the `latch` logic in the testing.

Reply:
Do you have an example? I'm not sure I got what you mean

Reply:
yes, exactly, it's only there for testing purpose.

Reply:
So then you can have your own implementation of `CountDownLatch` on tests, e.g. `AppLifecycleIntegrationTest` has that, or just search for `CountDownLatch` and `org.awaitility.kotlin.await`, both should give you ideas.

Reply:
dang, I was absolutely unaware about such thing as awaitility, thanks 👍  Refactored it to be like LifecycleWatcherTest